### PR TITLE
[Relay] [TOPI] `{relay,topi}.nn.sparse_transpose` for **Square** CSR matrices

### PR DESCRIPTION
**Pull request by yuluny2**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

Implementation of fast transpose for CSR matrices with time complexity of O(nnz(X) + N), where X is the input CSR matrix, nnz(X) is the number of nonzeros of X, and N is the length of one dimension of X.

Please refer to https://github.com/scipy/scipy/blob/v0.14.0/scipy/sparse/sparsetools/csr.h#L380 for the algorithm implemented in Scipy.

Note that only square CSR matrices are supported since support on the sparse shape has not been introduced to Relay, hence only the first dimension of input matrices can be inferred.

Future roadmap:
1. Support non-square matrices
2. Add schedules for different devices.
3. Support more sparse types

@ajtulloch @tmoreau89 I would really appraciate if you can review this PR.

Relay Test:
````
from tvm import relay
import topi
import tvm
from collections import namedtuple
import numpy as np
import scipy.sparse as sp

target = 'llvm'
data = relay.var("data")
indices = relay.var("indices")
indptr = relay.var("indptr")
CSR = namedtuple("CSR", ["data", "indices", "indptr"])
output = relay.nn.sparse_transpose(CSR(data, indices, indptr))

output_data, output_indices, output_indptr = output
# output is a relay tupletype

N = 1000
density = 0.5
x = sp.random(N, N, density=density, format='csr', dtype='float32')
params = {"data": x.data,
         "indices": x.indices,
         "indptr": x.indptr}

from tvm.contrib import graph_runtime
import time

func = relay.Function(relay.analysis.free_vars(output.astuple()), output.astuple())

print(func)

with relay.build_config(opt_level=0):
    graph, lib, params = relay.build(func, target, params=params)
    lib.save("lib.o")
ctx = tvm.context(target, 0)
m = graph_runtime.create(graph, lib, ctx)
m.set_input(**params)

print("finished compiling, testing one transpose time cost")
totaltime = 0
for i in range(30):
    st = time.time()
    # One transpose operation time
    m.run()
    end = time.time()
    # Retrieve output Tensor as numpy array
    outdata = m.get_output(0).asnumpy()
    outindices = m.get_output(1).asnumpy()
    outindptr = m.get_output(2).asnumpy()
    totaltime += (end-st)

output = sp.csr_matrix((outdata, outindices, outindptr), shape=(N,N)).todense()
print("One transpose operation time %f" % (totaltime/30))
assert np.allclose(output,x.todense().T)
````

### CODE REVIEW

File: src/relay/op/nn/sparse.cc [hunk 1]
```diff
@@ -93,5 +93,51 @@ RELAY_REGISTER_OP("nn.sparse_dense")
...
+
+
+RELAY_REGISTER_OP("nn.sparse_transpose")
+    .describe(R"code(Transpose a sparse matrix X. Only support square sparse matrix
+
+- **input**: `(N, N)`
+- **out**: `(N, N)`.
+
+)code" TVM_ADD_FILELINE)
+    .set_attrs_type_key("relay.attrs.SparseTransposeAttrs")
```
Comment (MarisaKirisame) [inline src/relay/op/nn/sparse.cc]: fix ident

File: topi/tests/python/test_topi_sparse.py [hunk 2]
```diff
@@ -235,6 +235,33 @@ def test_sparse_dense_csr():
     func(tvm.ndarray.array(X_np), tvm.ndarray.array(W_sp_np.data), tvm.ndarray.array(W_sp_np.indices), tvm.ndarray.array(W_sp_np.indptr), Y_tvm)
     tvm.testing.assert_allclose(Y_tvm.asnumpy(), Y_np, atol=1e-4, rtol=1e-4)
 
+def test_sparse_transpose_csr():
+    import scipy.sparse as sp
```
Comment (MarisaKirisame) [inline topi/tests/python/test_topi_sparse.py]: move this to top

File: src/relay/op/nn/sparse.cc [hunk 3]
```diff
@@ -93,5 +93,51 @@ RELAY_REGISTER_OP("nn.sparse_dense")
...
+    .set_num_inputs(3)
+    .add_argument("sparse_data", "1D Tensor", "Sparse data matrix.")
+    .add_argument("sparse_indices", "1D Tensor", "Sparse indices matrix.")
+    .add_argument("sparse_indptr", "1D Tensor", "Sparse indptr matrix.")
```
Comment (tmoreau89) [inline src/relay/op/nn/sparse.cc]: Spell out index pointer instead of indptr

File: topi/python/topi/nn/sparse.py [hunk 4]
```diff
@@ -101,3 +101,106 @@ def _compute_block(i, nb_j, j):
...
+def sparse_transpose(sparse_data, sparse_indices, sparse_indptr):
+    """
+    Transpose a square sparse matrix,
+    `A` is an n-by-n sparse matrix in the CSR format.
+    ** Currently only support Square Matrices **
+
+    Parameters
+    ----------
+    sparse_data : tvm.Tensor
+        1-D with shape [nonzeros], dtype of 'float32'
```
Comment (tmoreau89) [inline topi/python/topi/nn/sparse.py]: Does the data tensor have to be of type `float32` or can we play around with other datatypes?

File: src/relay/op/nn/sparse.cc (hunk 1 above)
Comment (Yulun-Yao) [reply src/relay/op/nn/sparse.cc]: applied

File: topi/tests/python/test_topi_sparse.py (hunk 2 above)
Comment (Yulun-Yao) [reply topi/tests/python/test_topi_sparse.py]: applied

File: topi/python/topi/nn/sparse.py (hunk 4 above)
Comment (Yulun-Yao) [reply topi/python/topi/nn/sparse.py]: Yes. We can support more data types in the future.

File: src/relay/op/nn/sparse.cc (hunk 3 above)
Comment (Yulun-Yao) [reply src/relay/op/nn/sparse.cc]: Resolved

**Comment by tmoreau89** [comment]
Thanks for resolving my comments. Let's wait for one more review, perhaps by @ajtulloch . Also now that there is traction on sparse operator support, we can perhaps discuss potential tensor types for sparse operators so we can correctly typecheck programs? Also solve the issue of enforcing the square matrix shape. This could be the object of an RFC.

**Comment by Yulun-Yao** [comment]
> Thanks for resolving my comments. Let's wait for one more review, perhaps by @ajtulloch . Also now that there is traction on sparse operator support, we can perhaps discuss potential tensor types for sparse operators so we can correctly typecheck programs? Also solve the issue of enforcing the square matrix shape. This could be the object of an RFC.

I don't think we can enforce matrix shape. If we can enforce matrix shape it means that we can directly infer shape from relay end.

**Comment by ajtulloch** [comment]
Looks good to me. Excited to see what you folks have planned :)

**Pull request merged**